lib/src/commit.rs

```diff
@@ -145,10 +146,18 @@ impl Commit {
         &self.data.description
     }
 
+    pub fn author(&self, mailmap: &Mailmap) -> Signature {
+        mailmap.resolve(&self.data.author)
+    }
```

Comment:
Is the mailmap thingy a backend or frontend/UI matter? (I don't follow the previous discussion, so I don't know the conclusion, sorry.)

If it's backend stuff, I think the raw signature can be (forcibly) translated by `GitBackend` (or `Store`) layer. There might be a global flag to temporary turn off the translation.

If it's frontend stuff, the translation should be made explicitly by templater, and I think the revset frontend can do reverse translation (so the revset engine impls don't have to deal with the mailmap.)

Reply:
It’s a frontend matter. You want canonical signatures to be used by default in the vast majority of contexts (hence exposing it as the “default” way to get this information out of a commit as a nudge), but there are use‐cases for inspecting and querying the raw information in the commits. Git offers a `--no-mailmap` option to do this, but offering it as separate DSL functions seemed more orthogonal and fitting with Jujutsu’s general interfaces. The translation also depends on the current working copy commit (to use the current `.mailmap` file), so it’s not trivial to do in a global way without plumbing through a lot more state.

Reply:
> It’s a frontend matter.

Thanks for clarification. In that case, I wouldn't add `Commit::author(mailmap)` functions, and just let templater/revset modules do translation.

> Git offers a `--no-mailmap` option to do this, but offering it as separate DSL functions seemed more orthogonal and fitting with Jujutsu’s general interfaces.

I'm not sure about the UI. I personally like opt-in interface (something like `author.mailmap().email()`, `author(mailmap:pattern)`), but that wouldn't be major opinion. If mailmap is enabled implicitly by default, and if it's unlikely to show both raw and translated signatures within the same session, adding `--no-mailmap` flag/config might be better than exposing `_raw()` DSL functions.

Reply:
I was thinking about tooling using the library that wants to display metadata about a commit without using the templating language. If it’s entirely the caller’s responsibility to realize they should be handling the `.mailmap`, and there’s nothing in the API nudging them towards that or even just pointing out that they have to manually do it, I fear it would result in what @nasamuffin described as the whack‐a‐mole experience that Git has had to deal with of ensuring that everything that wants to render a name/email maps it correctly. My hope is that by making it explicit in the interface when you should expect to need the `.mailmap` information, and clear what you’re doing when you explicitly request the “wrong” results, this could be avoided.

Of course, I guess the existing `jj` command‐line tool always uses templates to display this kind of thing. But it seems like we can’t really count on that for external tools, and it’s surprising that to get the correct canonical author information as reported by `jj` from the library you’d need to layer on additional non‐obvious APIs? A doc comment might help here, but is still weak against autocompleting the obvious function compared to getting the machine to tell you what you need and what your alternatives are.

My concern with making it opt‐in in templating functions is the same – customizing your template and putting `author()` in and getting an outdated representation of the commit author that disagrees with what Git tooling will give you seems like a foot‐gun. Adding an `--ignore-mailmap` option to every command that queries revsets or formats commits would be possible, but it seems like a lot of additional CLI surface churn and state plumbing compared to just exposing functions for that? If people really don’t like the additional functions, maybe the `*_raw` variants should just be removed and people can use lower‐level tools to poke at the raw object data if they want (or even just temporarily delete the `.mailmap` file).

Reply:
I agree with Yuya that it should be kept in the templater. A quick way to check how much of a problem it will be in practice is to find where we call .author() and .committer() today.

Reply:
> I was thinking about tooling using the library that wants to display metadata about a commit without using the templating language. If it’s entirely the caller’s responsibility to realize they should be handling the `.mailmap`, and there’s nothing in the API nudging them towards that or even just pointing out that they have to manually do it, I fear it would result in what @nasamuffin described as the whack‐a‐mole experience that Git has had to deal with of ensuring that everything that wants to render a name/email maps it correctly. My hope is that by making it explicit in the interface when you should expect to need the `.mailmap` information, and clear what you’re doing when you explicitly request the “wrong” results, this could be avoided.

I see your point. I think it's up to how much guarantee (and Git compatibility) we're going to provide. (I personally don't care about mailmap-ed identity, so I wouldn't mind if jj output is diverged from git.)

> Adding an `--ignore-mailmap` option to every command that queries revsets or formats commits would be possible, but it seems like a lot of additional CLI surface churn and state plumbing compared to just exposing functions for that?

I believe it's actually easier to implement. The resulting code would be something like
```rust
let mailmap = if global_args.ignore_mailmap {
    vec![]
} else {
    laod_mailmap(repo.get_commit(wc))?
};
...
RevsetContext {
   ...
   mailmap,
}
```
That said, I'm not sure about UI/UX things.

Reply:
I think it’s worth noting that consistent `.mailmap` support throughout tooling is a lot more important to some people than others. There are reasons for people to change name, like divorce or gender transition, where tooling unnecessarily surfacing their previous name can be both be upsetting to them and potentially confusing/misleading to others. It’s also important to people who want to accurately find commits by people who have changed their details, and there’s the case of people changing email due to things like loss of control over the account, where it would be a bad idea for people or automated systems to contact the old email because of its presence in a commit.

So, while I understand that to many people this feature will just be a triviality that they may not bother to use for their own information and that they feel they could take or leave for rendering, I think that the downside of not doing the right thing by default is a lot bigger than it might seem. Of course, we can’t solve this perfectly with the current data model, which bakes this data into commits permanently (hence this not resolving #2957), but I think it would be *really* unfortunate to do worse than Git here, and not just for reasons of compatibility for the sake of compatibility.

As far as the code structure goes – what about not exposing the mailmapping methods on `Commit`, but keeping the renames of the Rust API methods to `*_raw` (or `*_unmapped`?) and adding a doc comment explaining that the signature should be canonicalized before display or processing in most cases? That way the name will still signal to external users that there’s subtleties involved here, and hopefully make them inclined to check the documentation. The `jj` CLI doesn’t have many relevant calls here, but that’s because it extensively uses the templating language; I don’t think we can assume that every external tool that wants to get human‐suitable metadata out of commits will realize that they need to be indirectly going via the DSL, especially if the template function name stays the same as the Rust method name despite having different behaviour.

A global `--ignore-mailmap` option sounds okay to me, and I can implement it. It would mean that you can’t use the templating language to do things like compare raw signatures to what they’re getting mapped to, but that might not be a big deal.

Reply:
True, we want to make it easy for alternative UIs and such to get the mailmapped data, so I agree with you now that it should go in the lib crate. Thanks again for working on it. I'll try to find time to review it today.